main/src/ca/uwaterloo/flix/language/phase/Redundancy.scala

```diff
@@ -106,13 +106,40 @@ object Redundancy extends Phase[TypedAst.Root, TypedAst.Root] {
     }
 
     // Compute the used symbols inside the definition.
-    val usedExp = visitExp(defn.exp, Env.of(defn.fparams.map(_.sym)))
+    val recursionContext = RecursionContext(defn.sym, arity(defn))
+    val usedExp = visitExp(defn.exp, Env.of(Some(recursionContext)) ++ defn.fparams.map(_.sym))
 
     // Check for unused parameters and remove all variable symbols.
-    val usedAll = (usedExp ++ checkUnusedFormalParameters(usedExp) ++ checkUnusedTypeParameters(usedExp)).copy(varSyms = Set.empty)
+    val usedAll = (usedExp and checkUnusedFormalParameters(usedExp) and checkUnusedTypeParameters(usedExp)).copy(varSyms = Set.empty)
+
+    val usedAllWithUnconditionalRecursions = if (usedAll.unconditionallyRecurses) usedAll + UnconditionalRecursion(defn.sym) else usedAll
 
     // Check if the used symbols contains holes. If so, strip out all error messages.
-    if (usedAll.holeSyms.isEmpty) usedAll else usedAll.copy(errors = Set.empty)
+    if (usedAllWithUnconditionalRecursions.holeSyms.isEmpty)
+      usedAllWithUnconditionalRecursions
+    else
+      usedAllWithUnconditionalRecursions.copy(errors = Set.empty)
+  }
+
+  /**
+    * Finds the arity of the Def, including lambdas nested directly beneath it.
+    */
+  private def arity(defn: Def): Int = {
```

Comment:
Much better! But it would be even better to use defn.tpe to determine the arity. In general we try to rely on types when we can.